File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:    
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        if not lists:
            return None

        res_head = lists[0]
        if not res_head:
            return None

        arr = []

        for i in range(len(lists)):
            iter_head = lists[i]
            while iter_head:
                arr.append(iter_head)
                iter_head = iter_head.next

        arr = sorted(arr, key=lambda node: node.val)
        for i in range(len(arr) - 1):
            arr[i].next = arr[i + 1]

        return arr[0]
        


        

File: Data Structures & Algorithms/linked-list-cycle-detection/submission-0.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

public class Solution {
    public bool HasCycle(ListNode head) {
        var steps = 0;
        var trail = head;
        var ahead = head.next;

        while (ahead != null)
        {
            if (ahead == trail)
                return true;

            ahead = ahead.next;

            if (steps % 2 == 1)
                trail = trail.next;

            steps++;        
        }

        return false;
    }
}
